floris/optimization/load_optimization/load_optimization.py

```diff
@@ -0,0 +1,459 @@
+        max_dist_D (flat, optional): Maximum distance in rotor diameters. Defaults to 10.0.
+        exp_ws_std (float, optional): Exponent for the wind speed standard deviation.
+            Defaults to 1.0.
+        exp_thrust (float, optional): Exponent for the thrust. Defaults to 1.
+        initial_power_setpoint (np.array, optional): Initial power setpoint for each turbine.
+            Defaults to None.
+        derating_levels (np.array, optional): Array of derating levels to consider.
+            Defaults to np.linspace(1.0, 0.001, 5).
+
+    """
+
+    # Ensure we're in derating mode
+    fmodel.set_operation_model("simple-derating")
+
+    # If initial set point not provided, set to rated (assumed max) power
+    if initial_power_setpoint is None:
+        max_power = get_max_powers(fmodel)
+        initial_power_setpoint = np.tile(max_power, (fmodel.n_findex, 1))
+
+    # Initialize the the test power setpoints
+    test_power_setpoint = initial_power_setpoint.copy()
+    final_power_setpoint = initial_power_setpoint.copy()
+
+    # Get the sorted coords
+    sorted_indices = fmodel.core.grid.sorted_indices[:, :, 0, 0]
+
+    # Initialize the net revenue using the initial setpoints
+    fmodel.set(power_setpoints=initial_power_setpoint)
+    fmodel.run()
+    net_revenue = compute_net_revenue(
+        fmodel=fmodel,
+        A=A,
+        load_ambient_tis=load_ambient_tis,
+        wake_slope=wake_slope,
+        max_dist_D=max_dist_D,
+        exp_ws_std=exp_ws_std,
+        exp_thrust=exp_thrust,
+    )
+
+    # Now loop over turbines
+    for t in sorted_indices.T:
+        # Loop over derating levels
+        for d in derating_levels:
+            # Apply the proposed derating level to the test_power_setpoint matrix
+            test_power_setpoint[range(fmodel.n_findex), t] = (
+                initial_power_setpoint[range(fmodel.n_findex), t] * d
+            )
+
+            # Apply the setpoint to fmodel
+            fmodel.set(power_setpoints=test_power_setpoint)
+
+            # Run
+            fmodel.run()
+
+            # Get the net revenue
+            test_net_revenue = compute_net_revenue(
+                fmodel=fmodel,
+                A=A,
+                load_ambient_tis=load_ambient_tis,
+                wake_slope=wake_slope,
+                max_dist_D=max_dist_D,
+            )
+
+            # Get a map of where test_net_revenue is greater than net_revenue
+            update_mask = test_net_revenue > net_revenue
+
+            # Where update_mask is false, revert the test_power_setpoint to previous value
+            test_power_setpoint[~update_mask, :] = final_power_setpoint[~update_mask, :]
+
+            # Update the final_power_setpoint
+            final_power_setpoint[:, :] = test_power_setpoint[:, :]
+
+            # Update the net_revenue
+            net_revenue[update_mask] = test_net_revenue[update_mask]
+
+    # Return the final power setpoint
+    return final_power_setpoint
```

Comment:
I think it's common for optimization routines to return both x^* (the solution to the optimization problem, in this case the power setpoints) and v^* (the optimal cost/value). Perhaps we could also return the final value of `net_revenue`?

Reply:
I've added the optimal net revenue as an output (and renamed this function `optimize_power_setpoints`, which feels more explicit to me)